node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/nodeadmin/NodeAdminStateUpdater.java

```diff
@@ -161,4 +187,33 @@ private void adjustNodeAgentsToRunFromNodeRepository() {
                              .map(NodeSpec::getHostname)
                              .collect(Collectors.toList());
     }
+
+    private static class CachedSupplier<T> implements Supplier<T> {
```

Comment:
Alternatives to this is Guava's [Suppliers.memoizeWithExpiration](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoizeWithExpiration(com.google.common.base.Supplier,%20long,%20java.util.concurrent.TimeUnit)), but it does not support invalidation. There's also Guava's `Cache`, but that seems to produce even messier code...